GPII-2474:  Add support for "live" reloading of template content...

the-t-in-rtf opened the pull request:
See [GPII-2474](https://issues.gpii.net/browse/GPII-2474) for details.  I have assigned to @amb26 and added @sgithens as a reviewer, I will leave it to you to to involve other people or reassign as you see fit.

@sgithens, I have published a dev version that includes live support.  If you upgrade your `gpii-handlebars`dependency to `1.0.2-dev.20170629T094606Z.7194f07`, you should be able to try this out in your own work.  Note: you'll have to use the new `gpii.express.hb.live` class instead of `gpii.express.hb`.  See the docs for usage details.

the-t-in-rtf committed:
GPII-2474:  Add support for "live" reloading of template content.

amb26 commented on src/js/server/watcher.js:
```diff
@@ -0,0 +1,102 @@
+    members: {
+        watcher: null
+    },
+    eventsToWatch: ["add", "change", "unlink"], // Watch for file adds, changes, and removals.
```
This would more reusably be expressed as a hash of name to true/falsy, to make it easier for derivers from this grade to override listening to individual events

the-t-in-rtf replied:
Very good point.  Done, and fleshed out the docs with a clear example as well.

amb26 commented on tests/js/server/live-reload-tests.js:
```diff
@@ -0,0 +1,337 @@
+
+var gpii  = fluid.registerNamespace("gpii");
+
+require("../../../index");
```
fluid.require("%gpii-handlebars"), more readably?

the-t-in-rtf replied:
Done, throughout.  There were a bunch of variations on this, and plain `require` calls for `gpii-*` packages that I also fixed.

amb26 commented on src/js/server/watcher.js:
```diff
@@ -0,0 +1,102 @@
+        });
+    });
+
+    // fluid.log("Watching the following directories: ", resolvedDirs.join(","));
```
A few stray commented-out logs in this file - also missing parenthesis on line 33

the-t-in-rtf replied:
Fixed.  In trying to suss out which paren was left open, I also discovered one in `handlebars.md`, which is also now fixed.

amb26 commented on docs/watcher.md:
```diff
@@ -0,0 +1,43 @@
+# `gpii.handlebars.watcher`
+
+This component uses the ["chokidar" library](https://github.com/paulmillr/chokidar) to watch one or more directories
```
Better CHOKIDAR than ENKIDU

the-t-in-rtf replied:
Gilgamesh and Enkidu at Uruk.

amb26 commented on src/js/server/watcher.js:
```diff
@@ -0,0 +1,102 @@
+    Based on chokidar: https://github.com/paulmillr/chokidar
+
+ */
+// TODO: Discuss (with Antranig) moving this to its own micro-module if it's more broadly useful.
```
Sounds like a good idea, assuming that build tools eventually arise that make working with micromodules substantially less arduous : P

the-t-in-rtf replied:
Let's revisit as soon as we or anyone else has a second use case for the watcher grade.

the-t-in-rtf committed:
GPII-2474: Added a "reload" event to the inline template middleware, and added tests and docs to demonstrate live reloading of that content.

the-t-in-rtf commented:
@sgithens left [a comment on the associate JIRA ticket](https://issues.gpii.net/browse/GPII-2474?focusedCommentId=27025&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-27025) that reminded me that the client-side template bundles delivered by the "inline" middleware were not being reloaded.  I added the ability to reload that content, and added tests and documentation to demonstrate how to wire it up properly.

the-t-in-rtf committed:
GPII-2474: Updated watcher `eventsToWatch` to use a map rather than an array.

the-t-in-rtf committed:
GPII-2474: Removed unused `mock-fs` dependency left over from previous approach to "live reload" tests.

the-t-in-rtf committed:
GPII-2474: Remove stray comments and fixed unclosed parens in docs/comments.

the-t-in-rtf committed:
GPII-2474: Converted various `require` calls for this and other `gpii-*` packages to use `fluid.require`.

the-t-in-rtf committed:
GPII-2474: Fixed a timing problem that would result in the watcher attempting to scan directories that had been removed during test cleanup.

the-t-in-rtf commented:
While working on the feedback reported here and in the ticket, I started noticing ENOENT errors when running particular tests.  I found and fixed a timing bug in the tests, where cleanup was sometimes occurring while the watcher was still watching folders.

the-t-in-rtf committed:
GPII-2474: Updated various packages based on output of `npm outdated`.

the-t-in-rtf commented:
I have confirmed the new watcher approach works in Windows (via the Vagrant tests).  I also took the time to update dependencies.  @amb26, this should be ready for review again.

the-t-in-rtf committed:
NOJIRA: Removed TODO about previous test inconsistencies.

sgithens reviewed (commented):
Hi!  I just got a chance to try this out and it works great!  I just have 2 comments so far (having not gone through the actual PR code in detail).

1. It might be nice to bump down the reload time from 2 seconds, or if that is the default, maybe put in the documents the option that can be used to change.  I know at least for my system, it never takes that long to save to disk, and actually the trigger finger I developed from before this functionality existed using killed node, restarted node, and reloaded the browser in less than 2 seconds.

2. It's not a showstopper, but it would be cool if we could figure out a more inherited way to add the client-side reloading with adding an event/listener to the parent component, and the extra listeners to the gpii.express.hb.live.  I haven't looked under the hood in detail yet, but if any more of this could be made to happen just by switching the grade to gpii.express.hb.live that would be rad.

the-t-in-rtf commented:
Hi, @sgithens.  As we discussed last night, the timing of the reload polling is adjustable.  I'd like to start by changing the timing to be too short, i. e. so short that partial templates are loaded before the file is completely saved.  This is the problem the current settings aim to avoid, if I can get to a state where it's broken and tune from there, hopefully we can find a reasonable balance between speed and completeness of content.  In any case I will add docs about those specific settings.

Making client-side templates reload on the fly is a bit more complex.  The client-side component would have to poll the server-side fairly often.  If the "inline" middleware cleanly supported caching (etags, checking caching headers), the usual answer ("nothing has changed") is much lower cost.

Once we have the ability to detect updated templates via polling, we'd need more client-side infrastructure to identify content associated with an updated template and force a redraw.  I will give some thought to making a live TemplateAware grade.  Depending on how expensive the polling is, we may need to talk about how we ensure that live reloading is disabled by the time code reaches end users.

the-t-in-rtf committed:
GPII-2474: Added tests for large template writes and turned off `awaitWriteFinish`.  Expanded docs for that setting.

the-t-in-rtf commented:
So, I added a test that writes 25Mb of data to a new template.  I chose that number as it's more than twice the size of my largest template in a real project.

Rather than adjust the timing I started by simply disabling `awaitWriteFinish`.  There were no problems running the new test in either Windows or OS X, and from the timing of the template loading events and resulting test output, it looks like a reload occurs in less than 200ms now.

the-t-in-rtf commented:
I just committed the work to add caching support to the "inline" template middleware.  In addition to my own tests, I confirmed with an actual browser that a reload results in a 304 (not modified).

the-t-in-rtf committed:
GPII-2474: Added caching to the "inline" middleware.

the-t-in-rtf commented:
In my last conversation with @sgithens, he asked about creating convenience grades to simplify the wiring of client-side grades to "live" inline template middleware.

the-t-in-rtf committed:
GPII-2344: Updated vagrant config to remove drive mounting steps.

the-t-in-rtf committed:
GPII-2487: Initial update of dependencies to reduce vulnerabilities reported by snyk.

amb26 commented:
This branch looks good - the tests produce such a lot of junk in Temp though - eventually my disk filled up:

```
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:26 scoped_dir15916_29752
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:26 watcher-test-dir-ccgy3t28-748
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:26 scoped_dir15916_22343
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:26 scoped_dir15916_12346
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_13971
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_19248
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_29130
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_3598
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_20298
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_32764
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_32013
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_12802
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_23049
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_16596
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_5097
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_11724
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_757
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_15918
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_3677
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_2814
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_15290
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_23511
drwxrwx---+ 1 Bosmon None     0 Jul 25 00:27 scoped_dir15916_6617
```

amb26 commented:
Each scoped_dir is about 13MB

the-t-in-rtf commented:
I couldn't reproduce the cleanup issues in OS X.  I'm pushing the recent infusion fixes through to allow me to run the tests in Windows.  I just finished all of this package's dependencies and will continue troubleshooting on my next shift.

the-t-in-rtf committed:
GPII-2344: Updated dependencies to avoid problems running tests in Windows.

the-t-in-rtf commented:
This pull now includes the "root" fixes that would otherwise prevent this PR's tests from running in Vagrant.

the-t-in-rtf commented:
@amb26, I don't see the "scoped_dir" concerns in the VM.  Can you take a listing of the contents of one of them and share here?

amb26 commented:
I'm continuing to test this branch. In my vagrant Windows box, I'm not seeing the scoped_dirs - however, in all environments I am seeing intermittent and random test failures - perhaps about 1 run in 3. Here is a recent sample from my vagrant box - 

```
07:12:49.658:  jq: FAIL: Module "Tests for inlining of templates..." Test name "Confirm that caching works as expected..." - Message: The status code should indicate that the content hasn't changed... - at sequence position 1 of 12
07:12:49.658:  jq: Expected: 304
07:12:49.658:  jq: Actual: 200
07:12:49.658:  jq: Source:     at Object.assertEquals (C:\gpii-handlebars\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:137:19)
    at Request._callback (C:\gpii-handlebars\tests\js\server\inline-tests.js:39:20)
    at Request.self.callback (C:\gpii-handlebars\node_modules\request\request.js:188:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (C:\gpii-handlebars\node_modules\request\request.js:1171:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (C:\gpii-handlebars\node_modules\request\request.js:1091:12)
    at IncomingMessage.g (events.js:291:16)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
07:12:49.658:  jq: FAIL: Module "Tests for inlining of templates..." Test name "Confirm that caching works as expected..." - Message: The response should include a valid ETag header... - at sequence position 1 of 12
07:12:49.658:  jq: Expected: 98c72d5483a86f85a28241389e016eb9
07:12:49.658:  jq: Actual: 550a0d219f4d2ed0d849588eb57b668e
07:12:49.658:  jq: Source:     at Object.assertEquals (C:\gpii-handlebars\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:137:19)
    at Request._callback (C:\gpii-handlebars\tests\js\server\inline-tests.js:40:20)
    at Request.self.callback (C:\gpii-handlebars\node_modules\request\request.js:188:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (C:\gpii-handlebars\node_modules\request\request.js:1171:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (C:\gpii-handlebars\node_modules\request\request.js:1091:12)
    at IncomingMessage.g (events.js:291:16)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
07:12:49.674:  jq: FAIL: Module "Tests for inlining of templates..." Test name "Confirm that caching works as expected..." - Message: The body should be empty... - at sequence position 1 of 12
07:12:49.689:  jq: Expected: ""
07:12:49.689:  jq: Actual: {
  "ok": true,
  "templates": {
    "layouts": {
      "code": "{{{body}}}",
      "custom": "<h1>Layout</h1>\r\n\r\n<p>This is content coming from the layout.</p>\r\n\r\n<hr/>\r\n\r\n{{{body}}}\r\n\r\n",
      "initblock": "<html>\r\n  <head>\r\n      <title>Init Block Test Page...</title>\r\n      {{>common-includes}}\r\n\r\n      <div class=\"initBlock-viewport\"></div>\r\n\r\n      <script type=\"text/javascript\" src=\"/content/js/tests-initBlock.js\"></script>\r\n\r\n  </head>\r\n  <body>\r\n    {{{body}}}\r\n  </body>\r\n</html>\r\n",
      "main": "<h1>Main Layout</h1>\r\n\r\n<p>This is content coming from the layout.</p>\r\n\r\n<hr/>\r\n\r\n{{{body}}}\r\n\r\n",
      "overridden": "<h1>Layout</h1>\r\n\r\n<p>This is a layout found in the primary template directory.</p>\r\n\r\n<hr/>\r\n\r\n{{{body}}}\r\n\r\n"
    },
    "pages": {
      "code": "<p>If a code parameter was included in the path used in this request, it should appear below:</p>\r\n\r\n<p class=\"code\">{{req.params.code}}</p>\r\n",
      "custom-no-matching-layout": "<h2>Page</h2>\r\n\r\n<p>This is content coming from the page.</p>\r\n\r\n{{>partial}}\r\n\r\n\r\n",
      "custom": "<h2>Page</h2>\r\n\r\n<p>This is content coming from the page.</p>\r\n\r\n{{>partial}}\r\n\r\n\r\n",
      "error": "There was an error:\r\n\r\n{{message}}",
      "helpers-equals": "{{#equals 1 1}}good{{else}}bad{{/equals}}",
      "helpers-jsonify-custom-options": "{{{jsonify . stringifyStrings=true}}}\r\n",
      "helpers-jsonify-string": "{{{jsonify .}}}",
      "helpers-jsonify": "{{{jsonify json}}}\r\n",
      "helpers-md": "{{{md markdown}}}\r\n",
      "index": "<div>\r\n  <h2>Page</h2>\r\n\r\n  <p>This is content coming from the page.</p>\r\n\r\n  {{>partial}}\r\n</div>\r\n",
      "initblock": "{{{initBlock \"gpii.tests.initBlock\"}}}",
      "overridden": "<div>\r\n  <h2>Primary Override Page</h2>\r\n\r\n  <p>This is content coming from a page served up from the primary template directory.</p>\r\n\r\n  {{>overridden-partial}}\r\n</div>\r\n",
      "partials": "{{>renderer-partial}}",
      "replace": "<div class=\"viewport-replaceWith replaced\">\r\n    <h1>Content from the template:</h1>\r\n    <div>\r\n      {{{body}}}\r\n    </div>\r\n    <hr/>\r\n\r\n  {{>partial}}\r\n</div>",
      "singleTemplateMiddleware": "<p>This template exercises the singleTemplateMiddleware grade and confirms that it renders content and exposes request data as expected:</p>\r\n\r\n\r\n<h1>Request Data</h1>\r\n{{#if req.query.myvar}}<div id=\"req-myvar\">{{req.query.myvar}}</div>{{/if}}\r\n\r\n<h1>Query</h1>\r\n{{jsonify req.query}}",
      "variable": "Hello, {{variable}}.",
      "secondary": "<div>\r\n  <h2>Secondary Page</h2>\r\n\r\n  <p>This is content coming from a page served up from the secondary template directory.</p>\r\n\r\n  {{>secondary-partial}}\r\n</div>\r\n"
    },
    "partials": {
      "common-error": "{{#if message}}\r\n<div class=\"callout alert\">\r\n    Something went wrong.  Allow me to elaborate:\r\n\r\n    {{jsonify .}}\r\n</div>\r\n{{/if}}\r\n",
      "common-includes": "<!-- Bring in jQuery and other dependencies -->\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/lib/jquery/core/js/jquery.js\"></script>\r\n\r\n<!-- Bring in fluid dependencies -->\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/FluidDocument.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/Fluid.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/FluidDOMUtilities.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/FluidIoC.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/DataBinding.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/FluidView.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/lib/fastXmlPull/js/fastXmlPull.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/renderer/js/fluidParser.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/renderer/js/fluidRenderer.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/FluidRequests.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/ModelTransformation.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/infusion/src/framework/core/js/ModelTransformationTransforms.js\"></script>\r\n\r\n<!-- gpii-binder dependency -->\r\n<script type=\"text/javascript\" src=\"/modules/gpii-binder/src/js/binder.js\"></script>\r\n\r\n<!-- Bring in Handlebars dependency -->\r\n<script type=\"text/javascript\" src=\"/modules/handlebars/dist/handlebars.js\"></script>\r\n\r\n<!-- Bring in Pagedown dependencies -->\r\n<script type=\"text/javascript\" src=\"/modules/pagedown/Markdown.Converter.js\"></script>\r\n<script type=\"text/javascript\" src=\"/modules/pagedown/Markdown.Sanitizer.js\"></script>\r\n\r\n<!-- Bring in the components from this package -->\r\n<script type=\"text/javascript\" src=\"/src/js/common/helper.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/common/equals.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/common/jsonify.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/common/md-common.js\"></script>\r\n\r\n<script type=\"text/javascript\" src=\"/src/js/client/hasRequiredOptions.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/ajaxCapable.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/md-client.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/renderer.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/templateManager.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/templateAware.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/templateFormControl.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/templateMessage.js\"></script>\r\n<script type=\"text/javascript\" src=\"/src/js/client/templateRequestAndRender.js\"></script>\r\n\r\n",
      "common-success": "{{#if message}}\r\n<div class=\"callout success\">\r\n    {{message}}\r\n</div>\r\n{{/if}}\r\n",
      "form-contained-initial": "<div class=\"templateFormControl-success\"></div>\r\n<div class=\"templateFormControl-error\"></div>\r\n\r\nA place for everything, and everything in its place.\r\n<form></form>",
      "form-failure-initial": "<div class=\"templateFormControl-success\"></div>\r\n<div class=\"templateFormControl-error\"></div>\r\n<form>\r\n    <input type=\"submit\" class=\"submit\" value=\"{{buttonName}}\"/>\r\n</form>\r\n",
      "form-nested-initial": "<div class=\"outer-success\"></div>\r\n<div class=\"outer-error\"></div>\r\n<form class=\"outer-form\">\r\n    <input type=\"submit\" class=\"outer-submit\" value=\"{{buttonName}}\"/>\r\n</form>\r\n\r\n<div class=\"form-nested-nested\">This should not be visible</div>",
      "form-nested-nested": "<div class=\"inner-success\"></div>\r\n<div class=\"inner-error\"></div>\r\n<form class=\"inner-form\">\r\n    <input type=\"submit\" class=\"inner-submit\" value=\"{{buttonName}}\"/>\r\n</form>",
      "form-success-initial": "<div class=\"templateFormControl-success\"></div>\r\n<div class=\"templateFormControl-error\"></div>\r\n<form>\r\n    <input type=\"submit\" class=\"submit\" value=\"{{buttonName}}\"/>\r\n</form>\r\n",
      "form-success": "<div class=\"callout success\">\r\n    This was a triumph.  Let's celebrate with some moist, delicious data.\r\n\r\n  <hr/>\r\n\r\n  <div class=\"json\">{{jsonify .}}</div>\r\n</div>",
      "overridden-partial": "<p>Content from a partial served from the primary template directory:</p>\r\n\r\n{{>partial}}",
      "partial": "<p>Content from the partial:</p>\r\n\r\n<table>\r\n    <thead>\r\n    <tr>\r\n        <th>item</th><th>value</th>\r\n    </tr>\r\n    </thead>\r\n    <tbody>\r\n    <tr>\r\n        <td>model variable</td>\r\n        <td class=\"variable\">{{myvar}}</td>\r\n    </tr>\r\n    <tr>\r\n        <td>query variable</td>\r\n        <td class=\"query\">{{req.query.myvar}}</td>\r\n    </tr>\r\n    <tr>\r\n        <td>jsonify</td>\r\n        <td class=\"jsonify\">{{{jsonify json}}}</td>\r\n    </tr>\r\n    <tr>\r\n        <td>markdown</td>\r\n        <td class=\"markdown\">{{{md markdown}}}</td>\r\n    </tr>\r\n    <tr>\r\n        <td>Testing Equal Variables</td>\r\n        <td class=\"equal\">{{#equals json.baz json.qux}}true{{else}}false{{/equals}}</td>\r\n    </tr>\r\n    <tr>\r\n        <td>Testing Unequal Variables</td>\r\n        <td class=\"unequal\">{{#equals json.foo json.qux}}true{{else}}false{{/equals}}</td>\r\n    </tr>\r\n    <tr>\r\n        <td>Testing Equal Variable and String</td>\r\n        <td class=\"equal\">{{#equals json.foo \"bar\"}}true{{else}}false{{/equals}}</td>\r\n    </tr>\r\n    <tr>\r\n        <td>Testing Unequal Variable and String</td>\r\n        <td class=\"unequal\">{{#equals \"nonsense\" json.qux}}true{{else}}false{{/equals}}</td>\r\n    </tr>\r\n    </tbody>\r\n</table>",
      "renderer-partial": "This is partial content.",
      "secondary-partial": "<p>Content from a partial served from the secondary template directory:</p>\r\n\r\n{{>partial}}"
    }
  }
}
07:12:49.689:  jq: Source:     at Object.assertEquals (C:\gpii-handlebars\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:137:19)
    at Request._callback (C:\gpii-handlebars\tests\js\server\inline-tests.js:41:20)
    at Request.self.callback (C:\gpii-handlebars\node_modules\request\request.js:188:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (C:\gpii-handlebars\node_modules\request\request.js:1171:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (C:\gpii-handlebars\node_modules\request\request.js:1091:12)
    at IncomingMessage.g (events.js:291:16)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
07:12:49.689:  jq: Test concluded - Module "Tests for inlining of templates..." Test name "Confirm that caching works as expected...": 1/4 passed - FAIL
```

amb26 commented:
Updating my chromedriver in my physical box has eliminated the scoped_dir problems. However, in both environments there remains a watcher-test-dir containing a single 8-byte text file.

amb26 commented:
Test failure just received in my physical box (Win 7 x64) - I think this resembles a failure which Githens says he saw once

```
15:26:54.062:  FATAL ERROR: Uncaught exception: EPERM: operation not permitted, stat 'C:\Users\Bosmon\AppData\Local\Temp\live-templates-a33uzurn-7414'
Error: EPERM: operation not permitted, stat 'C:\Users\Bosmon\AppData\Local\Temp\live-templates-a33uzurn-7414'
    at Error (native)
15:26:54.064:  Current activity:
15:26:54.065:  jq: FAIL: Module "Testing live reloading of template content..." Test name "The 'inline' middleware should be able to handle reloads..." - Messag
e: Unexpected failure in test case (see following log for more details): EPERM: operation not permitted, stat 'C:\Users\Bosmon\AppData\Local\Temp\live-templates
-a33uzurn-7414'
15:26:54.066:  jq: Source:     at kettle.test.handleUncaughtException (E:\source\gits\gpii-handlebars\node_modules\kettle\lib\test\KettleTestUtils.js:31:15)
    at Object.fire (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1533:40)
    at process.onUncaughtException (E:\source\gits\gpii-handlebars\node_modules\infusion\src\module\fluid.js:101:31)
    at emitOne (events.js:96:13)
    at process.emit (events.js:188:7)
    at process._fatalException (bootstrap_node.js:296:26)
15:26:54.067:  FATAL ERROR: Uncaught exception: EPERM: operation not permitted, stat 'C:\Users\Bosmon\AppData\Local\Temp\live-templates-a33uzurn-7414'
Error: EPERM: operation not permitted, stat 'C:\Users\Bosmon\AppData\Local\Temp\live-templates-a33uzurn-7414'
    at Error (native)
15:26:54.068:  Current activity:
15:26:54.069:  jq: FAIL: Module "Testing live reloading of template content..." Test name "The 'inline' middleware should be able to handle reloads..." - Messag
e: Unexpected failure in test case (see following log for more details): EPERM: operation not permitted, stat 'C:\Users\Bosmon\AppData\Local\Temp\live-templates
-a33uzurn-7414'
15:26:54.070:  jq: Source:     at kettle.test.handleUncaughtException (E:\source\gits\gpii-handlebars\node_modules\kettle\lib\test\KettleTestUtils.js:31:15)
    at Object.fire (E:\source\gits\gpii-handlebars\node_modules\infusion\src\framework\core\js\Fluid.js:1533:40)
    at process.onUncaughtException (E:\source\gits\gpii-handlebars\node_modules\infusion\src\module\fluid.js:101:31)
    at emitOne (events.js:96:13)
    at process.emit (events.js:188:7)
    at process._fatalException (bootstrap_node.js:296:26)
15:26:54.075:  jq: Test concluded - Module "Testing live reloading of template content..." Test name "The 'inline' middleware should be able to handle reloads..
.": 2/4 passed - FAIL
```

the-t-in-rtf commented:
I have had no luck reproducing the EPERM errors, but I can at least reproduce the incorrect status code and etag errors in my BootCamp partition.  Will investigate that and the orphaned watcher content tomorrow.

the-t-in-rtf committed:
GPII-2474: Updated hamfisted caching test that used a hard-coded checksum.

the-t-in-rtf committed:
GPII-2474:  Removed additional hamfistedness (second hard-coded checksum).

the-t-in-rtf commented:
The "inline caching" checks were failing because I naively assumed that a key checksum would be the same across platforms.  Here's what the middleware does:

1. The middleware includes an etag in the initial response, based on a checksum of the stringified version of the current set of templates.
2. The client passes that etag in the headers of subsequent requests.
3. The middleware compares the client's etag value to the current checksum.
     a. If the checksums don't match, the template content (with the current etag) is returned.
     b. If the checksums match, the middleware indicates that the content has not changed with a "304" response.

Since I hard coded an expected checksum that didn't match what particular versions of Windows computed, we ended up at outcome 3a instead of 3b.  The likely cause is a difference in the actual template content (character encodings), i.e. the "canned" checksum was computed with one type of carriage return, and Windows used another.

I updated the test in question to retrieve the content once and use the "etag" header returned in a subsequent request, which is what we want to test anyway.  We're not comparing etags across instances running on different platforms, so it ultimately doesn't matter if one platform's etag for a given set of content doesn't match another.

the-t-in-rtf committed:
Clarified purpose of npm config step in `.vagrant.yml` file.

the-t-in-rtf committed:
GPII-2474: Improved cleanup in "watcher" tests.

the-t-in-rtf committed:
NOJIRA: Updated dependencies in preparation for further review.

the-t-in-rtf commented:
@amb26, I tracked down the "watcher" content that wasn't getting cleaned, updated all dependencies, and verified the tests are running consistently in Windows.  This should be ready for further review.  If at all possible, I'd like to merge this before starting in on the  unrelated spike to update to use "markdown-it".

the-t-in-rtf commented:
Thanks!  I cut the 1.0.2 release and updated the forward-facing version in master to 1.0.3.
